src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -10,12 +17,112 @@ namespace DB
 
 namespace ErrorCodes
 {
+extern const int CANNOT_WRITE_TO_KAFKA;
 extern const int MISSING_ACKNOWLEDGEMENT;
 extern const int INVALID_CONFIG_PARAMETER;
+extern const int TYPE_MISMATCH;
+}
+
+namespace KafkaStream
+{
+ChunkPartitioner::ChunkPartitioner(ContextPtr context, const Block & header, const ASTPtr & partitioning_expr_ast) {
+    /// `InterpreterCreateQuery::handleExternalStreamCreation` ensures this
+    assert(partitioning_expr_ast);
+
+    ASTPtr query = partitioning_expr_ast;
+    auto syntax_result = TreeRewriter(context).analyze(query, header.getNamesAndTypesList());
+    partitioning_expr = ExpressionAnalyzer(query, syntax_result, context).getActions(false);
+
+    partitioning_key_column_name = partitioning_expr_ast->getColumnName();
+
+    if (auto * shard_func = partitioning_expr_ast->as<ASTFunction>())
+        if (shard_func->name == "rand" || shard_func->name == "RAND")
+        {
+            random_partitioning = true;
+            std::random_device r;
```

Comment:
Is there dangling object here ? `rand` object ref a object `r()` which has been destructed ? We may need consider reusing pseudo generator as well.